File: python/src/agent_squad/agents/neutrosophic_supervisor.py
from agent_squad.agents.supervisor_agent import SupervisorAgent, SupervisorAgentOptions
from agent_squad.neutrosophic import decide, neutrosophic_consensus, score_text_response


class NeutrosophicSupervisor(SupervisorAgent):
    """Supervisor agent that adds neutrosophic consensus to team responses."""

    async def send_messages(self, messages: list[dict[str, str]]) -> str:
        """Process messages and return agent responses with neutrosophic consensus."""
        raw_response = await super().send_messages(messages)
        if not raw_response or raw_response.startswith("No agent matches"):
            return raw_response

        scored_responses = [
            (response, score_text_response(response))
            for response in self._split_agent_responses(raw_response)
        ]
        consensus = neutrosophic_consensus(score for _, score in scored_responses)
        action = decide(consensus)

        return self._format_neutrosophic_response(scored_responses, consensus, action.value)

    def _split_agent_responses(self, raw_response: str) -> list[str]:
        marker_positions: list[int] = []
        for agent in self.team:
            marker = f"{agent.name}: "
            start = raw_response.find(marker)
            while start != -1:
                marker_positions.append(start)
                start = raw_response.find(marker, start + len(marker))

        marker_positions = sorted(set(marker_positions))
        responses = []
        for index, start in enumerate(marker_positions):
            end = marker_positions[index + 1] if index + 1 < len(marker_positions) else len(raw_response)
            responses.append(raw_response[start:end].strip())

        return responses or [raw_response]

    @staticmethod
    def _format_neutrosophic_response(scored_responses, consensus, action: str) -> str:
        score_lines = [
            f"- {response}\n"
            f"  T={score.T:.2f}, I={score.I:.2f}, F={score.F:.2f}"
            for response, score in scored_responses
        ]

        return (
            "<neutrosophic_consensus>\n"
            f"action: {action}\n"
            f"T: {consensus.T:.2f}\n"
            f"I: {consensus.I:.2f}\n"
            f"F: {consensus.F:.2f}\n"
            "</neutrosophic_consensus>\n"
            "<agent_responses>\n"
            + "\n".join(score_lines)
            + "\n</agent_responses>"
        )

File: typescript/src/agents/neutrosophicSupervisor.ts
import { Agent } from "./agent";
import { SupervisorAgent, SupervisorAgentOptions } from "./supervisorAgent";
import {
  decide,
  neutrosophicEvidenceConsensus,
  scoreTextResponse,
  Triplet,
} from "../neutrosophic";

type SupervisorMessage = { recipient: string; content: string };
type ScoredResponse = { agentName: string; responseText: string; score: Triplet };

export class NeutrosophicSupervisor extends SupervisorAgent {
  constructor(options: SupervisorAgentOptions) {
    super(options);
  }

  protected async sendMessages(messages: SupervisorMessage[]): Promise<string> {
    this.validateMessages(messages);

    const tasks = messages
      .map((message) => {
        const agent = this.team.find((candidate) => candidate.name === message.recipient);
        return agent ? this.sendScoredMessage(agent, message.content) : null;
      })
      .filter((task): task is Promise<ScoredResponse> => task !== null);

    if (!tasks.length) {
      const consensus = new Triplet(0.0, 1.0, 0.0);
      return this.formatNeutrosophicResponse([], consensus, decide(consensus));
    }

    const scoredResponses = await Promise.all(tasks);
    const consensus = neutrosophicEvidenceConsensus(
      scoredResponses.map((response) => response.score)
    );

    return this.formatNeutrosophicResponse(
      scoredResponses,
      consensus,
      decide(consensus)
    );
  }

  private async sendScoredMessage(agent: Agent, content: string): Promise<ScoredResponse> {
    const rawResponse = await this.sendMessage(
      agent,
      content,
      this.userId,
      this.sessionId,
      this.additionalParams
    );
    const prefix = `${agent.name}: `;
    const responseText = rawResponse.startsWith(prefix)
      ? rawResponse.slice(prefix.length)
      : rawResponse;

    return {
      agentName: agent.name,
      responseText,
      score: scoreTextResponse(responseText),
    };
  }

  private validateMessages(messages: SupervisorMessage[]): void {
    if (!Array.isArray(messages)) {
      throw new TypeError("messages must be an array");
    }

    messages.forEach((message, index) => {
      if (typeof message !== "object" || message === null) {
        throw new TypeError(`messages[${index}] must be an object`);
      }
      if (typeof message.recipient !== "string" || !message.recipient.trim()) {
        throw new Error(`messages[${index}].recipient must be a non-empty string`);
      }
      if (typeof message.content !== "string" || !message.content.trim()) {
        throw new Error(`messages[${index}].content must be a non-empty string`);
      }
    });
  }

  private formatNeutrosophicResponse(
    scoredResponses: ScoredResponse[],
    consensus: Triplet,
    action: string
  ): string {
    const scoreLines = scoredResponses.map((response) => (
      `- ${response.agentName}: ${response.responseText}\n` +
      `  T=${response.score.T.toFixed(2)}, I=${response.score.I.toFixed(2)}, F=${response.score.F.toFixed(2)}`
    ));

    return (
      "<neutrosophic_consensus>\n" +
      `action: ${action}\n` +
      `T: ${consensus.T.toFixed(2)}\n` +
      `I: ${consensus.I.toFixed(2)}\n` +
      `F: ${consensus.F.toFixed(2)}\n` +
      "</neutrosophic_consensus>\n" +
      "<agent_responses>\n" +
      scoreLines.join("\n") +
      "\n</agent_responses>"
    );
  }
}
